Fix various bugs related to restart of StatefulSet pods that may result in connectivity issues

christarazi opened the pull request:
- **k8s/watcher: Remove outdated comments about K8sEventHandover**
- **daemon, endpoint: Consolidate K8s metadata into struct**
- **watchers: Detect Pod UID changes**
- **daemon,endpoint,cni: Pass pod UID through CNI ADD**
- **k8s/watchers: Fix pod IP modified check**

--- 

Fixes: https://github.com/cilium/cilium/issues/30409

christarazi committed:
daemon, endpoint: Consolidate K8s metadata into struct

Refactor the Kubernetes-fetched metadata into a consolidated struct type
for ease of adding a new type without increasing the function signatures
of the relevant functions.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
daemon,endpoint,cni: Pass pod UID through CNI ADD

Building off of the previous commit, This commit modifies if the CNI ADD
handling logic to pass the Kubernetes Pod UID in the ADD request to the
Agent. In addition, a extra condition checks whether the UID of the pod
corresponding to the CNI ADD request is the same UID which Cilium is
aware of (within its pod informer store) during endpoint creation. This
is key for handling StatefulSets properly, see [1].

If the UID is outdated, this means that Cilium has not yet received the
updated pod event that corresponds with the CNI ADD. In this case,
Cilium will attempt to query the pod store up for to 2 seconds. If the
pod store does not have the altest pod reference after 2 seconds, Cilium
will trigger a direct fetch from the apiserver, as to avoid a
potentially large window of time where an endpoint cannot be created. If
the direct fetch fails, the endpoint will be created with the 'init'
identity until the metadata resolver (controller) is able to fetch the
latest pod metadata.

The reason for these changes is due to the way StatefulSets are
implemented in Kubernetes. One significant property of StatefulSets is
that they have a fixed name. When a StatefulSet is restarted, Cilium is
unable to properly handle it because Cilium indexes based on
namespace/name[1]. Therefore, we must pass a UID through the CNI ADD in
order to ensure that when Cilium creates a new endpoint, it is created
with the most up-to-date corresponding Kubernetes metadata.

[1]: https://github.com/cilium/cilium/issues/30409, see case 1.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
watchers: Detect Pod UID changes

This commit causes the Kubernetes Pod watcher to trigger endpoint
identity updates if the Pod UID changes. This covers the third scenario
pointed out in [1], see issue for more details.

To summarize, with StatefulSet restarts, it is possible for the
apiserver to coalesce Pod events where the delete and add event are
replaced with a single update event that includes, for example a label
change. If Cilium indexes based on namespace/name of the pod, then it
will miss this update event and therefore, Cilium must use the UID of
the pod to disambiguate.

[1]: https://github.com/cilium/cilium/issues/30409

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi commented on pkg/identity/identity.go:
```diff
@@ -64,6 +64,7 @@ type IPIdentityPair struct {
 	Metadata     string          `json:"Metadata"`
 	K8sNamespace string          `json:"K8sNamespace,omitempty"`
 	K8sPodName   string          `json:"K8sPodName,omitempty"`
+	K8sUID       string          `json:"K8sUID,omitempty"`
```
This is a change to the API model of the kvstore interaction. However, given that it's an additive change, it should be fully JSON backwards/forwards compat.

tklauser commented on pkg/endpoint/endpoint.go:
```diff
@@ -292,6 +292,10 @@ type Endpoint struct {
 
+	// K8sUID is the Kubernetes UID of the pod. Passed directly from the CNI.
+	// Immutable after Endpoint creation.
+	K8sUID string
```
Given this is immutable after endpoint creation, should this be unexported?

christarazi replied:
Don't we need to exported for marshaling purposes? As in `json.{Unm,M}arshal` requires fields to be exported for them to get detected.

tklauser replied:
Yeah, good point. I wasn't considering this struct is marshalled.

tklauser commented on pkg/endpoint/api.go:
```diff
@@ -76,6 +76,7 @@ func NewEndpointFromChangeModel(ctx context.Context, owner regeneration.Owner, p
 	ep.dockerEndpointID = base.DockerEndpointID
 	ep.K8sPodName = base.K8sPodName
 	ep.K8sNamespace = base.K8sNamespace
+	ep.K8sUID = base.K8sUID
```
Do we need to consider this field during endpoint restore serialization/deserialization in endpoint (i.e. `(*Endpoint).toSerializedEndpoint` and `(*Endpoint).ep.fromSerializedEndpoint`) as well?

aanm replied:
Yes!

christarazi replied:
Thanks, good catch!

marseel replied:
~From what I see we currently don't use UID after endpoint restoration, right? Just wanted to confirm my understanding that currently in this PR we only use UID during endpoint creation AFAIK?~

tklauser reviewed (approved):
Thanks for the descriptive commit messages!

LGTM with some non-blocking questions inline.

aanm commented on pkg/k8s/watchers/pod.go:
```diff
@@ -400,7 +405,9 @@ func (k *K8sWatcher) updateK8sPodV1(oldK8sPod, newK8sPod *slim_corev1.Pod) error
-		if labelsChanged {
+		if labelsChanged || uidChanged {
+			// Consider a UID change the same as a label change in case the pod's
+			// identity needs to be updated, see GH-30409.
```
Please leave comments on why it's not necessary to do the same for when the annotations change.

christarazi replied:
I'll update the comment. The short story is that annotations don't affect identities, so we don't need to trigger the logic for updating labels if the annotations changed.

aanm commented on daemon/cmd/endpoint.go:
```diff
@@ -453,6 +455,13 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
+				logfields.K8sUID:     ep.K8sUID,
+			}).Warn("Detected outdated Pod UID during Endpoint creation. " +
+				"The Endpoint may be created with outdated identity due to delay in Kubernetes event updates, i.e. until Pod event is processed by Cilium. " +
+				"If the Endpoint persists with an outdated identity for an extended period of time, please report an issue.")
```
I'm not sure if we should Warn in here, this is a transient error and we will recover afterwards.

christarazi replied:
Wouldn't it be useful to know that the bugs we've hit with statefulsets were detected and Cilium is warning the user that there might be a period of time where the identity is out of date? It seems useful to me.

marseel replied:
+1 from me for warning, I think it's useful. 

Another way to prevent that would be to fail CNI ADD when we don't have a pod in the informer store and let Kubelet retry later on, but this has lots of other consequences. Or wait for a fixed amount of time for store to get updated pod and if it doesn't happen then fail CNI ADD. But this is probably quite a lot of work and outside of this PR's scope.

marseel replied:
Another place similar to this is metadataResolver https://github.com/cilium/cilium/blob/1518818397ac98be01573bb0c52cd0ece122f3f7/pkg/endpoint/endpoint.go#L1776
where we could get a "random" pod instead of the correct one in case the store is lagging.

How about an alternative approach:
- we try to get the pod from the store
- if it's not there or has incorrect UID we fetch it from k8s control plane directly like in high-scale case. Not ideal, but in most cases, it should not be triggered and it would ensure we have the correct pod information always.

christarazi replied:
While you are correct, I don't see any negative consequences for having an out-of-date pod object reference in the `Endpoint`. In fact, we don't even update the reference on pod update events, so they are pretty much always stale after `metadataResolver` runs. This is because we don't use the metadata for anything else besides initially fetching the labels for identity purposes. Label changes are detected in the pod update events, but the reference to the object is never updated.

To be clear, the only "negative consequence" is that it will have the same behavior that the warning log msg mentions: the endpoint's identity will be outdated until a pod update event comes in.

Is there something else I'm missing?

marseel replied:
> I don't see any negative consequences for having an out-of-date pod object reference

If we have reference to previous pod with incorrect UID then in endpoint synchronizer we will set ownerReference to incorrect pod that no longer exists:
https://github.com/cilium/cilium/blob/02d04b34005c45d955a6850d3faf561d9eae8f78/pkg/endpointmanager/endpointsynchronizer.go#L115
which means, kube-controller-manager will GC that CEP immediately.

christarazi replied:
I've changed the code to fetch the latest pod object in the case that the UID is mismatching during endpoint creation.

squeed replied:
How will the user know that a pod has been successfully updated? Should we log a message indicating that a pod was stale for XXX minutes?

Also, error messages "aren't real", only alerts are. There should be a metric for whether or not an endpoint has a correct pod.

christarazi replied:
There will be log msgs indicating what Cilium is doing when the outdated UID case is hit. We suspect this case to be very rare anyway.

aanm commented on pkg/k8s/watchers/pod.go:
```diff
@@ -539,14 +539,29 @@ func (k *K8sWatcher) deleteK8sPodV1(pod *slim_corev1.Pod) error {
+	//
+	// This specific scenario is to guard against a delayed delete event. See
+	// GH-30409.
+
```
I'm lost, you left this comment here but don't we need some code as well? :confused:

christarazi replied:
I've clarified the comment to be a note to readers/developers who might add new code below this comment in the future to keep this case in mind.

aanm commented on clustermesh-apiserver/clustermesh/root.go:
```diff
@@ -277,6 +277,11 @@ func (es *endpointSynchronizer) upsert(ctx context.Context, key resource.Key, ob
+					// Assume the first ownerRef is the pod.
+					entry.K8sUID = string(endpoint.OwnerReferences[0].UID)
+				}
+
```
No, we should check the ownerreference type.

tklauser reviewed (approved):
> @aanm @tklauser Here's the [diff](https://github.com/cilium/cilium/compare/13d3ac71eb2ba1d4018d6d7307880428a652dd7a..3eeb5d3bfe85570e296fb661631c1e31c74a0047) for implementing your feedback.

LGTM, thank you Chris!

marseel reviewed (commented):
Clustermesh changes look good, but I will take a look at the whole PR if you don't mind Chris, before approving :)

marseel commented on pkg/endpoint/endpoint.go:
```diff
@@ -1308,6 +1312,14 @@ func (e *Endpoint) GetK8sNamespace() string {
 
+// GetK8sUID returns the UID of the pod if the endpoint represents a Kubernetes
+// pod.
+func (e *Endpoint) GetK8sUID() string {
```
nit: this is not used anywhere, right? I would probably remove it.

christarazi replied:
Now it's used in the below fixup commit

marseel commented on pkg/k8s/watchers/cilium_endpoint.go:
```diff
@@ -116,13 +116,13 @@ func (k *K8sWatcher) endpointUpdated(oldEndpoint, endpoint *types.CiliumEndpoint
 				}
 				if !v4Added {
-					portsChanged := k.ipcache.DeleteOnMetadataMatch(oldPair.IPV4, source.CustomResource, endpoint.Namespace, endpoint.Name)
+					portsChanged := k.ipcache.DeleteOnMetadataMatch(oldPair.IPV4, source.CustomResource, endpoint.Namespace, endpoint.Name, string(endpoint.UID))
```
This is incorrect UID, right?
Seems to me like it's CiliumEndpoint UID and not pod UID/ ownerReference UID.

christarazi replied:
Good point, I'll fix this up.

christarazi replied:
<details><summary>Diff</summary>
<p>

```diff
commit b6be38f8e5de748c852376f87a2737d09a83855d
Author: Chris Tarazi <chris@isovalent.com>
Date:   Tue Apr 9 14:02:34 2024 -0700

    fixup! ipcache,watchers,identity: Pass pod UID into ipcache metadata
    
    Signed-off-by: Chris Tarazi <chris@isovalent.com>

diff --git a/api/v1/models/ip_list_entry_metadata.go b/api/v1/models/ip_list_entry_metadata.go
index 75a0099252..6efafe32e9 100644
--- a/api/v1/models/ip_list_entry_metadata.go
+++ b/api/v1/models/ip_list_entry_metadata.go
@@ -29,6 +29,9 @@ type IPListEntryMetadata struct {
 	// Source of the IP entry and its metadata
 	// Example: k8s
 	Source string `json:"source,omitempty"`
+
+	// UID of the backing resource to the IP (e.g. Kubernetes pod UID)
+	UID string `json:"uid,omitempty"`
 }
 
 // Validate validates this IP list entry metadata
diff --git a/api/v1/openapi.yaml b/api/v1/openapi.yaml
index 3a773d5ceb..ca93a92e1b 100644
--- a/api/v1/openapi.yaml
+++ b/api/v1/openapi.yaml
@@ -3787,6 +3787,9 @@ definitions:
       name:
         description: Name assigned to the IP (e.g. Kubernetes pod name)
         type: string
+      uid:
+        description: UID of the backing resource to the IP (e.g. Kubernetes pod UID)
+        type: string
   NamedPorts:
     description: |-
       List of named Layer 4 port and protocol pairs which will be used in Network
diff --git a/api/v1/server/embedded_spec.go b/api/v1/server/embedded_spec.go
index 7314db4792..37e3b02204 100644
--- a/api/v1/server/embedded_spec.go
+++ b/api/v1/server/embedded_spec.go
@@ -3573,6 +3573,10 @@ func init() {
           "description": "Source of the IP entry and its metadata",
           "type": "string",
           "example": "k8s"
+        },
+        "uid": {
+          "description": "UID of the backing resource to the IP (e.g. Kubernetes pod UID)",
+          "type": "string"
         }
       }
     },
@@ -9423,6 +9427,10 @@ func init() {
           "description": "Source of the IP entry and its metadata",
           "type": "string",
           "example": "k8s"
+        },
+        "uid": {
+          "description": "UID of the backing resource to the IP (e.g. Kubernetes pod UID)",
+          "type": "string"
         }
       }
     },
diff --git a/daemon/cmd/ipcache.go b/daemon/cmd/ipcache.go
index 31d85eac9a..9fdf84792a 100644
--- a/daemon/cmd/ipcache.go
+++ b/daemon/cmd/ipcache.go
@@ -86,6 +86,7 @@ func (ipc *ipCacheDumpListener) OnIPIdentityCacheChange(modType ipcache.CacheMod
 			Source:    string(newID.Source),
 			Namespace: k8sMeta.Namespace,
 			Name:      k8sMeta.PodName,
+			UID:       k8sMeta.UID,
 			// TODO (jrajahalme): Consider if named ports should be
 			//                    made visible in the model.
 		}
diff --git a/pkg/k8s/factory_functions.go b/pkg/k8s/factory_functions.go
index 46eec43c94..42ef3a411f 100644
--- a/pkg/k8s/factory_functions.go
+++ b/pkg/k8s/factory_functions.go
@@ -457,8 +457,9 @@ func TransformToCiliumEndpoint(obj interface{}) (interface{}, error) {
 				ResourceVersion: concreteObj.ObjectMeta.ResourceVersion,
 				// We don't need to store labels nor annotations because
 				// they are not used by the CEP handlers.
-				Labels:      nil,
-				Annotations: nil,
+				Labels:          nil,
+				Annotations:     nil,
+				OwnerReferences: slim_metav1.SlimOwnerReferences(concreteObj.OwnerReferences),
 			},
 			Encryption: func() *cilium_v2.EncryptionSpec {
 				enc := concreteObj.Status.Encryption
@@ -492,8 +493,9 @@ func TransformToCiliumEndpoint(obj interface{}) (interface{}, error) {
 					ResourceVersion: ciliumEndpoint.ObjectMeta.ResourceVersion,
 					// We don't need to store labels nor annotations because
 					// they are not used by the CEP handlers.
-					Labels:      nil,
-					Annotations: nil,
+					Labels:          nil,
+					Annotations:     nil,
+					OwnerReferences: slim_metav1.SlimOwnerReferences(ciliumEndpoint.ObjectMeta.OwnerReferences),
 				},
 				Encryption: func() *cilium_v2.EncryptionSpec {
 					enc := ciliumEndpoint.Status.Encryption
diff --git a/pkg/k8s/watchers/cilium_endpoint.go b/pkg/k8s/watchers/cilium_endpoint.go
index 5e62d3273e..7cc44c1288 100644
--- a/pkg/k8s/watchers/cilium_endpoint.go
+++ b/pkg/k8s/watchers/cilium_endpoint.go
@@ -167,10 +167,11 @@ func (k *K8sWatcher) endpointUpdated(oldEndpoint, endpoint *types.CiliumEndpoint
 		return
 	}
 
+	podUID := k.fetchPodUIDForCEP(endpoint)
 	k8sMeta := &ipcache.K8sMetadata{
 		Namespace:  endpoint.Namespace,
 		PodName:    endpoint.Name,
-		UID:        string(endpoint.UID),
+		UID:        podUID,
 		NamedPorts: make(ciliumTypes.NamedPortMap, len(endpoint.NamedPorts)),
 	}
 	for _, port := range endpoint.NamedPorts {
@@ -206,18 +207,19 @@ func (k *K8sWatcher) endpointUpdated(oldEndpoint, endpoint *types.CiliumEndpoint
 }
 
 func (k *K8sWatcher) endpointDeleted(endpoint *types.CiliumEndpoint) {
+	podUID := k.fetchPodUIDForCEP(endpoint)
 	if endpoint.Networking != nil {
 		namedPortsChanged := false
 		for _, pair := range endpoint.Networking.Addressing {
 			if pair.IPV4 != "" {
-				portsChanged := k.ipcache.DeleteOnMetadataMatch(pair.IPV4, source.CustomResource, endpoint.Namespace, endpoint.Name, string(endpoint.UID))
+				portsChanged := k.ipcache.DeleteOnMetadataMatch(pair.IPV4, source.CustomResource, endpoint.Namespace, endpoint.Name, podUID)
 				if portsChanged {
 					namedPortsChanged = true
 				}
 			}
 
 			if pair.IPV6 != "" {
-				portsChanged := k.ipcache.DeleteOnMetadataMatch(pair.IPV6, source.CustomResource, endpoint.Namespace, endpoint.Name, string(endpoint.UID))
+				portsChanged := k.ipcache.DeleteOnMetadataMatch(pair.IPV6, source.CustomResource, endpoint.Namespace, endpoint.Name, podUID)
 				if portsChanged {
 					namedPortsChanged = true
 				}
@@ -228,3 +230,20 @@ func (k *K8sWatcher) endpointDeleted(endpoint *types.CiliumEndpoint) {
 		}
 	}
 }
+
+func (k *K8sWatcher) fetchPodUIDForCEP(endpoint *types.CiliumEndpoint) string {
+	// Fetch the Pod UID from the internal Endpoint object if it exists,
+	// otherwise check the ownerRef of the CEP object.
+	var podUID string
+	ep := k.endpointManager.LookupCEPName(endpoint.Namespace + "/" + endpoint.Name)
+	if ep != nil {
+		// If the endpoint is local to this node, then this should suffice.
+		podUID = ep.GetK8sUID()
+	}
+	if podUID == "" {
+		if len(endpoint.OwnerReferences) > 0 && endpoint.OwnerReferences[0].Kind == "Pod" {
+			podUID = string(endpoint.OwnerReferences[0].UID)
+		}
+	}
+	return podUID
+}
```

</p>
</details>

marseel replied:
You've updated it in endpointDelete, but endpointUpdate is still having incorrect endpoint.uid, right?

christarazi replied:
I'm not sure I understand what else `endpointUpdate` needs? It's already been fixed to use the pod UID.

christarazi replied:
I've updated the `defer` function to replace the remaining reference to `endpoint.UID`.

marseel commented on pkg/ipcache/ipcache.go:
```diff
@@ -759,11 +761,11 @@ func (ipc *IPCache) GetNamedPorts() (npm types.NamedPortMultiMap) {
 	defer ipc.mutex.Unlock()
 	k8sMeta := ipc.getK8sMetadata(IP)
-	if k8sMeta != nil && k8sMeta.Namespace == namespace && k8sMeta.PodName == name {
+	if k8sMeta != nil && k8sMeta.Namespace == namespace && k8sMeta.PodName == name && k8sMeta.UID == uid {
```
Isn't `k8sMeta.UID` going to be always empty for restored endpoints after upgrade :thinking: ?

christarazi replied:
It will be and so this check will be true, so it shouldn't be an issue. When a new pod is created while running the new code, then the UID will be non-empty. Either the UID is set or it's empty, in other words, there shouldn't be a situation where it's mixed. So I don't see an issue, do you have something in mind?

marseel replied:
okay, I was thinking about deletion of pod that was previously created with versions that didn't support UID:
- Pod is running with the previous version of Cilium that doesn't handle UID - so when we restore endpoint k8s it won't have UID
- We upgrade it to version from this PR that now handles UID
- new cilium-agent restores the endpoint without UID in k8sMeta
- I guess we receive endpoint update event during the initialization and update k8s metadata? This is open question, I haven't checked code in depth, but looking at https://github.com/cilium/cilium/blob/dbb579ada86ebcde8bf41b10a55b23528099d4f8/pkg/ipcache/ipcache.go#L332 we might not be updating k8sMeta if identity did not change?
- endpoint is deleted and call to `DeleteOnMetadataMatch` has uid in it, so we need to make sure k8sMeta.UID is set after upgrade (previous point that I haven't fully checked). If it isn't then we would not trigger delete in ipcache even though we should.

christarazi replied:
One approach we can take here is to update the endpointsynchronizer `sync-to-k8s-ciliumendpoint` controller to also set the pod UID in the `Endpoint` object similar to what's being done already for CEP UID. However, there's no guarantee that this controller runs before [the ipache code above](https://github.com/cilium/cilium/blob/dbb579ada86ebcde8bf41b10a55b23528099d4f8/pkg/ipcache/ipcache.go#L332) is invoked. I'm not sure of another way around this than to skip the UID check if the UID is empty in [that ipcache code](https://github.com/cilium/cilium/blob/dbb579ada86ebcde8bf41b10a55b23528099d4f8/pkg/ipcache/ipcache.go#L332). What do you think?

marseel replied:
> I'm not sure of another way around this than to skip the UID check if the UID is empty

yea, I was thinking we could do something like that and then in 1.17 we could assume that UID is always non-empty. That would be probably the safest approach.

christarazi replied:
https://github.com/cilium/cilium/pull/31605#issuecomment-2097483504

marseel commented on pkg/ipcache/ipcache.go:
```diff
@@ -759,11 +761,11 @@ func (ipc *IPCache) GetNamedPorts() (npm types.NamedPortMultiMap) {
 // DeleteOnMetadataMatch removes the provided IP to security identity mapping from the IPCache
 // if the metadata cache holds the same "owner" metadata as the triggering pod event.
-func (ipc *IPCache) DeleteOnMetadataMatch(IP string, source source.Source, namespace, name string) (namedPortsChanged bool) {
+func (ipc *IPCache) DeleteOnMetadataMatch(IP string, source source.Source, namespace, name, uid string) (namedPortsChanged bool) {
```
Commenting on random like, this refers to line 304:
We probably want to add UID to `scopedLog`:
```
			scopedLog = scopedLog.WithFields(logrus.Fields{
				logfields.K8sPodName:   k8sMeta.PodName,
				logfields.K8sNamespace: k8sMeta.Namespace,
				logfields.NamedPorts:   k8sMeta.NamedPorts,
			})
```

marseel reviewed (changes requested):
Huge thanks @christarazi for working on this, looks good overall.

I am a bit worried about the upgrade path though, so I left a few comments.

marseel commented on pkg/k8s/watchers/cilium_endpoint.go:
```diff
@@ -227,3 +230,20 @@ func (k *K8sWatcher) endpointDeleted(endpoint *types.CiliumEndpoint) {
 }
+
+func (k *K8sWatcher) fetchPodUIDForCEP(endpoint *types.CiliumEndpoint) string {
+	// Fetch the Pod UID from the internal Endpoint object if it exists,
```
I find it a bit counter-intuitive that we are mixing object information received from k8s control plane and locally cached endpointManager. 

I would probably reverse order and first try ownerReference as a more reasonable/reliable source of truth, which will work for every CEP as long as it's managed by Cilium

christarazi replied:
Fair enough, I can switch it to just the ownerReference.

learnitall commented on daemon/cmd/endpoint.go:
```diff
@@ -455,6 +457,13 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
+				logfields.K8sUID:     ep.K8sUID,
+			}).Warn("Detected outdated Pod UID during Endpoint creation. " +
+				"The Endpoint may be created with outdated identity due to delay in Kubernetes event updates, i.e. until Pod event is processed by Cilium. " +
+				"If the Endpoint persists with an outdated identity for an extended period of time, please report an issue.")
```
[This question is related to https://github.com/cilium/cilium/pull/31605/commits/7dd6b4c697cb253634bdc4847c209c6c4d419f8f#r1550314914 but I'm starting it in another thread as to not interrupt the conversation surrounding handling the case of outdated pod UID]

I think the log message is a bit ambiguous in a couple of ways, so I have some clarifying questions:

1. What is "an extended period of time"? One minute? Five minutes?
2. Would users have to keep in mind the identity change grace period when determining how long it took for an Endpoint's identity to update after seeing this message?
3. Is the delay in Kubernetes event updates normal or expected? I feel this log message could be a great indicator when an API server is backed up with events and needs some attention, but the message tells users to only take action based off of the endpoint identity update delay.

christarazi replied:
See https://github.com/cilium/cilium/pull/31605#issuecomment-2097483504

learnitall commented on pkg/redirectpolicy/manager_test.go:
```diff
@@ -182,9 +182,11 @@ var (
+		ServiceID: k8s.ServiceID{
+			Name:      pod1.Name,
+			Namespace: pod1.Namespace,
+		},
```
Can we add a test here to check for a regression in the behavior that this commit fixes?

christarazi replied:
While I'm not opposed to adding tests, I'm afraid that in order to fully test out the scenarios within https://github.com/cilium/cilium/issues/30409, it would not suffice in a single package unit test nor would it be easy to write an e2e test for it, which requires figuring out a set of steps to repro the conditions. The alternative is to write a controlplane test, but I'm also afraid that that framework is not quite up to spec to support the testing needed, i.e. creating endpoints, simulating delays in k8s events, etc. and would be non-trivial to do so. I'm hesitant to block the PR on this. I also understand that it's not ideal to have a PR without tests either. :/

On the plus side, the changes are relatively easy to understand especially regarding the LRP code, which essentially is updated to query an internal map with a more unique key, instead.

learnitall replied:
My worry is that if we let the PR through without tests they will never be added. I feel that a separate TODO item for adding tests would go stale or become de-prioritized and never get done. I agree that the changes are relatively easy to understand, and delaying the PR isn't great, but I think it's a better alternative to merging these changes without the associated regression tests. This is a great opportunity to expand on the controlplane test framework to support testing these kinds of changes.

christarazi replied:
I've greatly simplified the PR as per https://github.com/cilium/cilium/pull/31605#issuecomment-2097483504, so this code highlighted here no longer exists.

christarazi replied:
After syncing offline, given the multi month effort investment required to allow the controlplane framework to have the ability to test this kind of behavior, I am resolving this thread to proceed with the PR as it resolves problematic behavior for users today.

learnitall commented on clustermesh-apiserver/clustermesh/root.go:
```diff
@@ -277,6 +277,10 @@ func (es *endpointSynchronizer) upsert(ctx context.Context, key resource.Key, ob
 					entry.Key = uint8(endpoint.Encryption.Key)
 				}
 
+				if len(endpoint.OwnerReferences) > 0 && endpoint.OwnerReferences[0].Kind == "Pod" {
```
Why do we only check the first owner reference and not iterate through the slice to look for a Pod owner?

christarazi replied:
The ownerRef on a CEP is only expected to be a Pod. There aren't multiple ownerRefs.

learnitall reviewed (commented):
Thank you for the helpful commit descriptions and comments! I have a couple of clarifying questions. My main concern here is that the PR doesn't include any tests which look for regressions in this behavior. It would be awesome if new go-based tests could be added to check for each of the scenarios laid out in #30409.

christarazi committed:
k8s/watcher: Remove outdated comments about K8sEventHandover

K8sEventHandover is no longer a flag so remove any remaining references
to it.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

squeed commented on pkg/redirectpolicy/manager.go:
```diff
@@ -46,8 +47,12 @@ type svcCache interface {
+type podID = struct {
+	k8s.ServiceID
+
+	uid types.UID
```
Careful here. If a pod is deleted and added with the same name, it is possible for the shared informer to coalesce this in to a single update.

As a general rule, unless we really need to associate with the specific "version" of a pod, we should ignore UIDs.

christarazi replied:
https://github.com/cilium/cilium/pull/31605#issuecomment-2097483504

squeed commented on pkg/identity/identity.go:
```diff
@@ -64,6 +64,7 @@ type IPIdentityPair struct {
 	Metadata     string          `json:"Metadata"`
 	K8sNamespace string          `json:"K8sNamespace,omitempty"`
 	K8sPodName   string          `json:"K8sPodName,omitempty"`
+	K8sUID       string          `json:"K8sUID,omitempty"`
```
Why do we need to persist the pod UID in the kvstore? What cases does it disambiguate?

christarazi replied:
https://github.com/cilium/cilium/pull/31605#issuecomment-2097483504

squeed commented on pkg/ipcache/ipcache.go:
```diff
@@ -759,11 +762,11 @@ func (ipc *IPCache) GetNamedPorts() (npm types.NamedPortMultiMap) {
 // DeleteOnMetadataMatch removes the provided IP to security identity mapping from the IPCache
 // if the metadata cache holds the same "owner" metadata as the triggering pod event.
-func (ipc *IPCache) DeleteOnMetadataMatch(IP string, source source.Source, namespace, name string) (namedPortsChanged bool) {
+func (ipc *IPCache) DeleteOnMetadataMatch(IP string, source source.Source, namespace, name, uid string) (namedPortsChanged bool) {
```
In what cases do we need UID? I'm having a bit of trouble thinking through all the cases.

Separately, I'm concerned that there are valid state transitions that this might break. e.g., if we restart a watch, and a CEP had transitioned from pod A to pod A'. We would never see the delete of A and the add of A', just the update. Would stale ipcache entries then be left behind?

christarazi replied:
https://github.com/cilium/cilium/pull/31605#issuecomment-2097483504

aanm commented on clustermesh-apiserver/clustermesh/root.go:
```diff
@@ -277,6 +277,10 @@ func (es *endpointSynchronizer) upsert(ctx context.Context, key resource.Key, ob
+				if len(endpoint.OwnerReferences) > 0 && endpoint.OwnerReferences[0].Kind == "Pod" {
+					entry.K8sUID = string(endpoint.OwnerReferences[0].UID)
+				}
+
```
What I meant here was that we should find the first occurrence of the "Pod". The pod can be on index 2 for example.

aanm commented on pkg/endpointmanager/endpointsynchronizer.go:
```diff
@@ -202,6 +205,8 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
 
 						scopedLog.WithField(logfields.CEPUID, localCEP.UID).Debug("storing CEP UID after create")
 						e.SetCiliumEndpointUID(localCEP.UID)
+						// TODO: Set here as well
```
What's the TODO?

aanm commented on pkg/endpointmanager/endpointsynchronizer.go:
```diff
@@ -161,13 +161,16 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
 						}
+						if err := updatePodUID(scopedLog, e, localCEP.GetUID()); err != nil {
+							return err
+						}
```
Wait a second, why are we updating the pod UID with the endpoint's UID?

christarazi replied:
Yeah, this is a mistake. I'll update.

aanm commented on pkg/endpointmanager/endpointsynchronizer.go:
```diff
@@ -202,6 +205,8 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
 						scopedLog.WithField(logfields.CEPUID, localCEP.UID).Debug("storing CEP UID after create")
 						e.SetCiliumEndpointUID(localCEP.UID)
+						// TODO: Set here as well
+						e.K8sUID = string(localCEP.UID)
```
Wait a second, why are we updating the pod UID with the endpoint's UID?

aanm commented on pkg/endpointmanager/endpointsynchronizer.go:
```diff
@@ -228,14 +233,16 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
 							return err
 						}
-
+						if err := updatePodUID(scopedLog, e, localCEP.GetUID()); err != nil {
```
Wait a second, why are we updating the pod UID with the endpoint's UID?

aanm commented on pkg/ipcache/ipcache.go:
```diff
@@ -74,6 +74,8 @@ type K8sMetadata struct {
 	// PodName is the Kubernetes pod name behind the IP
 	PodName string
+	// UID is the Kubernetes UID of the pod
+	UID string
```
Let's make the variable more obvious and change it to `PodUID`

aanm commented on pkg/ipcache/ipcache.go:
```diff
@@ -759,11 +762,17 @@ func (ipc *IPCache) GetNamedPorts() (npm types.NamedPortMultiMap) {
 // DeleteOnMetadataMatch removes the provided IP to security identity mapping from the IPCache
 // if the metadata cache holds the same "owner" metadata as the triggering pod event.
-func (ipc *IPCache) DeleteOnMetadataMatch(IP string, source source.Source, namespace, name string) (namedPortsChanged bool) {
+func (ipc *IPCache) DeleteOnMetadataMatch(IP string, source source.Source, namespace, name, uid string) (namedPortsChanged bool) {
```
Can we change the varilable name to be podUID?

aanm commented on pkg/ipcache/kvstore.go:
```diff
@@ -97,7 +97,7 @@ func newKVReferenceCounter(s store) *kvReferenceCounter {
 // kvstore, which will subsequently trigger an event in NewIPIdentityWatcher().
 func UpsertIPToKVStore(ctx context.Context, IP, hostIP netip.Addr, ID identity.NumericIdentity, key uint8,
-	metadata, k8sNamespace, k8sPodName string, npm types.NamedPortMap) error {
+	metadata, k8sNamespace, k8sPodName, k8sUID string, npm types.NamedPortMap) error {
```
same here with `k8sPodUID`

aanm commented on pkg/ipcache/kvstore.go:
```diff
@@ -120,6 +120,7 @@ func UpsertIPToKVStore(ctx context.Context, IP, hostIP netip.Addr, ID identity.N
 		Key:          key,
 		K8sNamespace: k8sNamespace,
 		K8sPodName:   k8sPodName,
+		K8sUID:       k8sUID,
```
same here with `k8sPodUID`

aanm commented on pkg/k8s/watchers/cilium_endpoint.go:
```diff
@@ -227,3 +230,12 @@ func (k *K8sWatcher) endpointDeleted(endpoint *types.CiliumEndpoint) {
 	}
 }
+
+func (k *K8sWatcher) fetchPodUIDForCEP(endpoint *types.CiliumEndpoint) string {
```
This doesn't need to be a method of  `(k *K8sWatcher)`, it can be a normal function.

aanm commented on pkg/k8s/watchers/cilium_endpoint.go:
```diff
@@ -227,3 +230,12 @@ func (k *K8sWatcher) endpointDeleted(endpoint *types.CiliumEndpoint) {
+	// otherwise check the ownerRef of the CEP object.
+	if len(endpoint.OwnerReferences) > 0 && endpoint.OwnerReferences[0].Kind == "Pod" {
+		return string(endpoint.OwnerReferences[0].UID)
+	}
```
The same comment I left with regarding the pod not being in the index 0 should be applied here as well.

aanm commented on pkg/redirectpolicy/manager.go:
```diff
@@ -46,8 +47,12 @@ type svcCache interface {
+type podID = struct {
+	k8s.ServiceID
+
+	uid types.UID
```
same rename from `uid` to `podUID`

squeed commented on pkg/k8s/watchers/pod.go:
```diff
@@ -352,12 +352,17 @@ func (k *K8sWatcher) updateK8sPodV1(oldK8sPod, newK8sPod *slim_corev1.Pod) error
 	oldPodIPLen := len(oldK8sPod.Status.PodIP)
 	newPodIPLen := len(newK8sPod.Status.PodIP)
 	switch {
+	case uidChanged:
```
This seems wrong; pod IPs and labels can change, even when a pod isn't deleted and recreated.

I see below that the pod IP equality check is wrong; that needs to be fixed as part of this PR.

christarazi replied:
Fixed in recent revisions

squeed commented on pkg/k8s/watchers/pod.go:
```diff
@@ -397,7 +402,10 @@ func (k *K8sWatcher) updateK8sPodV1(oldK8sPod, newK8sPod *slim_corev1.Pod) error
 
 	for _, podEP := range podEPs {
-		if labelsChanged {
+		if labelsChanged || uidChanged {
```
What case does this catch? That the CEP's labels are different from the Pod's labels?

If so, we should fix that regardless of the UID.

christarazi replied:
This catches scenario 3 from the linked issue in the PR description.

squeed commented on pkg/k8s/watchers/pod.go:
```diff
@@ -347,12 +344,17 @@ func (k *K8sWatcher) updateK8sPodV1(oldK8sPod, newK8sPod *slim_corev1.Pod) error
 	labelsChanged := !maps.Equal(oldPodLabels, newPodLabels)
+	uidChanged := oldK8sPod.UID != newK8sPod.UID
 
 	lrpNeedsReassign := false
```
```suggestion
	lrpNeedsReassign := !maps.Equal(oldPodLabels, newPodLabels) || !k8sTypes.IPSlice(oldPodIPs).DeepEqual(newPodIPs)
	// delete switch statement.
```

Everything that can change **with** a uid change, can change **without** one too :-(

christarazi replied:
IIUC, this essentially means `lrpNeedsReassign := true`, right? I'm not sure I understand why each pod event means we have to trigger the LRP logic if there could be no label or IP change. The current logic already detects if the labels or IPs have changed.

Given that the modification I made just adds UID to list of conditions, it isn't making any fundmental changes to the logic. If the suggestion here is to fix that core logic, I'd opt to have that be in another PR so we can get this one merged for fixing statefulset handling.

squeed replied:
Sorry, I made a typo. Corrected the comment.

squeed replied:
The problem is that we don't catch IP address changes; the code as-written just checks to see if the length changes. Which is silly; pods can change IPs at will. If we correctly catch IP and label changes, then there's no need to check UIDs.

christarazi replied:
Gotcha, I can implement this suggestion as a separate commit. I'll leave in the UID check as another OR as per scenario 3 in https://github.com/cilium/cilium/issues/30409.

christarazi committed:
k8s/watchers: Fix pod IP modified check

Previously, the logic in the pod watcher did not properly check for
changes in a pod's IPs, incorrectly relying on the length of the IP
strings changing, rather than the contents. This commit fixes that while
also simplifying the logic into a simpler one-liner.

Reported-by: Casey Callendrello <cdc@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

squeed commented on daemon/cmd/endpoint.go:
```diff
@@ -421,37 +436,68 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
+
+			t, ok := d.endpointCreations.requests[d.endpointCreations.getKey(ep)]
+			if !ok || time.Now().Sub(t.started) < 30*time.Second {
+				return ep, 0, fmt.Errorf("pod store out-of-date")
```
So we return failure here, and wait for kubelet -> containerd -> cni -> agent to retry? Am I reading this right?

Does it make sense to pause for a few seconds first to let the informer catch up? That would save some unnecessary system load.

christarazi replied:
Ok, I've pushed this suggestion, PTAL.

marseel reviewed (approved):
Thanks, looks good from my point of view.
I would probably add retry with a few seconds delay that Casey suggested.

squeed commented on daemon/cmd/endpoint.go:
```diff
@@ -559,6 +590,33 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
+	// Average attempt every 10ms.
+	err = resiliency.Retry(ctx, 2*time.Second, 12, func(_ context.Context, _ int) (bool, error) {
+		pod, k8sMetadata, err = d.fetchK8sMetadataForEndpoint(ep.K8sNamespace, ep.K8sPodName)
+		if pod != nil && ep.K8sUID != string(pod.GetUID()) {
```
One minor suggestion: handle the case where the CNI-supplied uid is missing. Just in case.

otherwise LGTM.

squeed reviewed (approved):
One minor suggestion, then LGTM

learnitall reviewed (approved):
API changes LGTM!
